File: _snippets/physics.py
data = {

    "Physics:增加对船的冲突碰撞 （不会被撞退）": {
        "prefix": "Physics:CollidesWith",
        "body": "Physics:CollidesWith(${1:COLLISION.OBSTACLES})",
        "description": "增加对船的冲突碰撞 （不会被撞退）"
    },
    "Physics:休眠状态下物理碰撞依旧生效": {
        "prefix": "Physics:SetDontRemoveOnSleep",
        "body": "Physics:SetDontRemoveOnSleep(${1:true})",
        "description": "休眠状态下物理碰撞依旧生效"
    },
    "Physics:关闭所有物理性(物理碰撞机全面暂停)": {
        "prefix": "Physics:SetActive",
        "body": "Physics:SetActive(${1:false})",
        "description": "关闭所有物理性(物理碰撞机全面暂停)"
    },
    "Physics:逻辑判断物理机是否处于工作状态": {
        "prefix": "Physics:IsActive",
        "body": "Physics:IsActive(${1:})",
        "description": "逻辑判断物理机是否处于工作状态"
    },
    "Physics:自由落体速率 0为不掉落，默认1质量": {
        "prefix": "Physics:SetMass",
        "body": "Physics:SetMass(${1:0})",
        "description": "自由落体速率 0为不掉落，默认1质量"
    },
    "Physics:物理范围": {
        "prefix": "Physics:SetSphere",
        "body": "Physics:SetSphere(${1:.5})",
        "description": "物理范围"
    },
    "Physics:设置摩擦": {
        "prefix": "Physics:SetFriction",
        "body": "Physics:SetFriction(${1:.1})",
        "description": "设置摩擦"
    },
    "Physics:设置阻尼 对速度的衰弱": {
        "prefix": "Physics:SetDamping",
        "body": "Physics:SetDamping(${1:0})",
        "description": "设置阻尼 对速度的衰弱"
    },
    "Physics:设置恢复 落地反弹力 反作用力": {
        "prefix": "Physics:SetRestitution",
        "body": "Physics:SetRestitution(${1:.5})",
        "description": "设置恢复 落地反弹力 反作用力"
    },
    "Physics:设置碰撞组": {
        "prefix": "Physics:SetCollisionGroup",
        "body": "Physics:SetCollisionGroup(${1:COLLISION.ITEMS})",
        "description": "设置碰撞组"
    },
    "Physics:清除冲突面": {
        "prefix": "Physics:ClearCollisionMask",
        "body": "Physics:ClearCollisionMask(${1:})",
        "description": "清除冲突面"
    },
    "Physics:物理碰撞体积，碰撞高度": {
        "prefix": "Physics:SetCapsule",
        "body": "Physics:SetCapsule(${1:rad},${2: 1})",
        "description": "物理碰撞体积，碰撞高度"
    },
    "Physics:中空类型的物理性": {
        "prefix": "Physics:SetCylinder",
        "body": "Physics:SetCylinder(${1:.6},${2: 2})",
        "description": "中空类型的物理性"
    },
    "Physics:设置三角形网格": {
        "prefix": "Physics:SetTriangleMesh",
        "body": "Physics:SetTriangleMesh(${1:PLAYER_COLLISION_MESH})",
        "description": "设置三角形网格"
    },
    "Physics:休眠时，物理特性不删除": {
        "prefix": "Physics:SetDontRemoveOnSleep",
        "body": "Physics:SetDontRemoveOnSleep(${1:true})",
        "description": "休眠时，物理特性不删除"
    },
    "Physics:设置碰撞关闭， 设定坐标后再开启碰撞": {
        "prefix": "Physics:SetCollides",
        "body": "Physics:SetCollides(${1:false})",
        "description": "设置碰撞关闭， 设定坐标后再开启碰撞"
    },
    "Physics:将自己约束到目标实体身上": {
        "prefix": "Physics:ConstrainTo",
        "body": "Physics:ConstrainTo(${1:target.entity})",
        "description": "将自己约束到目标实体身上"
    },
    "Physics:获取质量": {
        "prefix": "Physics:GetMass",
        "body": "Physics:GetMass(${1:})",
        "description": "获取质量"
    },
    "Physics:获取物理碰撞范围": {
        "prefix": "Physics:GetRadius",
        "body": "Physics:GetRadius(${1:})",
        "description": "获取物理碰撞范围"
    },
    "Physics:物理传送": {
        "prefix": "Physics:Teleport",
        "body": "Physics:Teleport(${1:x},${2:y},${3:z})",
        "description": "物理传送"
    },
    "Physics:清除转换历史记录": {
        "prefix": "Physics:ClearTransformationHistory",
        "body": "Physics:ClearTransformationHistory(${1:})",
        "description": "清除转换历史记录"
    },
    "Physics:获取速度向量": {
        "prefix": "Physics:GetVelocity",
        "body": "Physics:GetVelocity(${1:})",
        "description": "获取速度向量"
    },
    "Physics:物理运行覆盖": {
        "prefix": "Physics:SetMotorVelOverride",
        "body": "Physics:SetMotorVelOverride(${1:0},${2:0},${3:0})",
        "description": "物理运行覆盖"
    },
    "Physics:清除运行": {
        "prefix": "Physics:ClearMotorVelOverride",
        "body": "Physics:ClearMotorVelOverride(${1:})",
        "description": "清除运行"
    },
    "Physics:设置碰撞回调函数": {
        "prefix": "Physics:SetCollisionCallback",
        "body": "Physics:SetCollisionCallback(${1:OnCollide})",
        "description": "设置碰撞回调函数"
    },
    "Physics:获取当前速度": {
        "prefix": "Physics:GetMotorSpeed",
        "body": "Physics:GetMotorSpeed(${1:})",
        "description": "获取当前速度"
    },
    "Physics:获取当前三个方向上的速度分量": {
        "prefix": "Physics:GetMotorVel",
        "body": "Physics:GetMotorVel(${1:})",
        "description": "获取当前三个方向上的速度分量"
    },
}

File: scripts/system_components/Physics.lua
return {
    ["CheckGridOffset"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["ClearCollidesWith"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["ClearCollisionMask"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "清除冲突面",
        ["params"] = {},
        ["returns"] = {}
    },
    ["ClearLocalCollisionMask"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["ClearMotorVelOverride"] = {
        ["author"] = "lan",
        ["tips"] = "清除物理运行覆盖",
        ["params"] = {},
        ["returns"] = {}
    },
    ["ClearTransformationHistory"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "清除转换历史记录",
        ["params"] = {},
        ["returns"] = {}
    },
    ["CollidesWith"] = {
        ["author"] = "lan",
        ["tips"] = "增加碰撞(不会被撞退)",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "", ["param"] = "COLLISION"}
        },
        ["returns"] = {},
        replace_body = "Physics:CollidesWith(${1:COLLISION.OBSTACLES})",
    },
    ["ConstrainTo"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "将自己约束到目标实体身上",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "", ["param"] = "target.entity"}
        },
        ["returns"] = {}
    },
    ["GeoProbe"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["GetCollisionGroup"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["GetCollisionMask"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["GetHeight"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["GetMass"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "获取质量",
        ["params"] = {},
        ["returns"] = {}
    },
    ["GetMotorSpeed"] = {
        ["author"] = "lan",
        ["tips"] = "获取当前速度",
        ["params"] = {},
        ["returns"] = {{["explain"] = "", ["type"] = "num"}}
    },
    ["GetMotorVel"] = {
        ["author"] = "lan",
        ["tips"] = "获取当前三个方向上的速度分量",
        ["params"] = {},
        ["returns"] = {
            {["explain"] = "v_x", ["type"] = "num"},
            {["explain"] = "v_y", ["type"] = "num"},
            {["explain"] = "v_z", ["type"] = "num"}
        }
    },
    ["GetRadius"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "获取物理碰撞范围",
        ["params"] = {},
        ["returns"] = {}
    },
    ["GetVelocity"] = {
        ["author"] = "lan",
        ["tips"] = "获取速度向量",
        ["params"] = {},
        ["returns"] = {
            {["explain"] = "vx", ["type"] = "num"},
            {["explain"] = "vy", ["type"] = "num"},
            {["explain"] = "vz", ["type"] = "num"}
        }
    },
    ["IsActive"] = {
        ["author"] = "lan",
        ["tips"] = "逻辑判断物理机是否处于工作状态",
        ["params"] = {},
        ["returns"] = {}
    },
    ["IsPassable"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["SetActive"] = {
        ["author"] = "lan",
        ["tips"] = "关闭所有物理性(物理碰撞机全面暂停)",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "bool", ["param"] = "false"}
        },
        ["returns"] = {}
    },
    ["SetCapsule"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "物理碰撞体积 碰撞高度",
        ["params"] = {
            {["explain"] = "", ["type"] = "num", ["param"] = "rad"},

            {["explain"] = "", ["type"] = "num", ["param"] = "1"}
        },
        ["returns"] = {}
    },
    ["SetCollides"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "设置碰撞关闭 设定坐标后再开启碰撞",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "bool", ["param"] = "false"}
        },
        ["returns"] = {}
    },
    ["SetCollisionCallback"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "设置碰撞回调函数",
        ["params"] = {
            {
                ["explain"] = "",
                ["type"] = "fn",
                ["param"] = "fn",
                fn_params = {
                    {
                        ["explain"] = "碰撞者自己",
                        ["type"] = "ent",
                        ["param"] = "inst"
                    },
                    {
                        ["explain"] = "被碰撞者",
                        ["type"] = "ent",
                        ["param"] = "other"
                    }
                }
            }
        },
        ["returns"] = {},
        replace_body = "Physics:SetCollisionCallback(function(inst,other)\\n\\t$1\\nend)\\n",
    },
    ["SetCollisionGroup"] = {
        ["author"] = "",
        ["tips"] = "设置碰撞组",
        ["params"] = {
            [1] = {
                ["explain"] = "",
                ["type"] = "",
                ["param"] = "COLLISION.ITEMS"
            }
        },
        ["returns"] = {}
    },
    ["SetCollisionMask"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["SetCylinder"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "中空类型的物理性",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "num", ["param"] = "radius"},

            [2] = {["explain"] = "", ["type"] = "num", ["param"] = "3"}
        },
        ["returns"] = {}
    },
    ["SetDamping"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "设置阻尼 对速度的衰弱",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "num", ["param"] = "0"}
        },
        ["returns"] = {}
    },
    ["SetDontRemoveOnSleep"] = {
        ["author"] = "lan",
        ["tips"] = "休眠状态下物理碰撞依旧生效",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "bool", ["param"] = "true"}
        },
        ["returns"] = {}
    },
    ["SetFriction"] = {
        ["author"] = "lan",
        ["tips"] = "设置摩擦",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "num", ["param"] = "F"}
        },
        ["returns"] = {}
    },
    ["SetLocalCollisionMask"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["SetMass"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "自由落体速率 0为不掉落，默认1质量",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "num", ["param"] = "0"}
        },
        ["returns"] = {}
    },
    ["SetMotorVel"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["SetMotorVelOverride"] = {
        ["author"] = "lan",
        ["tips"] = "物理运行覆盖",
        ["params"] = {
            {["explain"] = "", ["type"] = "num", ["param"] = "vx"},
            {["explain"] = "", ["type"] = "num", ["param"] = "vy"},
            {["explain"] = "", ["type"] = "num", ["param"] = "vz"}
        },
        ["returns"] = {}
    },
    ["SetRestitution"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "设置恢复 落地反弹力 反作用力",
        ["params"] = {
            [1] = {["explain"] = "", ["type"] = "num", ["param"] = ".5"}
        },
        ["returns"] = {}
    },
    ["SetSphere"] = {
        ["author"] = "lan 球体",
        ["tips"] = "物理范围",
        ["params"] = {
            [1] = {["explain"] = "半径", ["type"] = "num", ["param"] = "r"}
        },
        ["returns"] = {}
    },
    ["SetTempMass0"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["SetTriangleMesh"] = {
        ["author"] = "不知名笔记",
        ["tips"] = "设置三角形网格",
        ["params"] = {
            [1] = {
                ["explain"] = "",
                ["type"] = "",
                ["param"] = "PLAYER_COLLISION_MESH"
            }
        },
        ["returns"] = {}
    },
    ["SetVel"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["Stop"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["TEMPHACK_DisableSleepDeactivation"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["Teleport"] = {
        ["author"] = "lan",
        ["tips"] = "物理传送",
        ["params"] = {
            {["explain"] = "", ["type"] = "", ["param"] = "x"},
            {["explain"] = "", ["type"] = "", ["param"] = "y"},
            {["explain"] = "", ["type"] = "", ["param"] = "z"}

        },
        ["returns"] = {}
    },
    ["TeleportOffset"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    },
    ["TeleportRespectingInterpolation"] = {
        ["author"] = "",
        ["tips"] = "UNKNOWN",
        ["params"] = {[1] = {["explain"] = "", ["type"] = "", ["param"] = ""}},
        ["returns"] = {}
    }

}
